File: Algorithms/Sorting_and_Searching/binarySearch.py
"""Binary Search. Theta(logn)."""

def binarySearch(array, start, end, x):
    
    if start > end:
        return -1
    
    middle = (start + end) // 2
    if x == array[middle]:
        return middle

    if x < array[middle]:
        return binarySearch(array, start, middle - 1, x)
    else:
        return binarySearch(array, middle + 1, end, x)

    
if __name__ == '__main__':
    array = [1, 21, 31, 43, 54]
    x = 54
    
    l = 0
    r = len(array) - 1
    result = binarySearch(array, l, r, x)
    
    if result != -1:
        print("Element is present at index ", result)
    else:
        print("Element is not present in the array")

File: Algorithms/Sorting_and_Searching/binarySearch.cpp
// Binary Search. Theta(log(n)).
#include <bits/stdc++.h>
using namespace std;

int binarySearch(int arr[], int l, int r, int x){
	if (l <= r){
		int m = l + (r - l) / 2;

		if (arr[m] == x){
			return m;
		}

		if (arr[m] > x){
			return binarySearch(arr, l, m - 1, x);
		} 
	        if (arr[m] < x){
			return binarySearch(arr, m + 1, r, x);
		}
	}
	return -1; // element has not been found
}

int main(void){
	int arr[] = { 2, 3, 4, 6, 7, 12, 34, 75, 89 };
	int x = 12;
	int n = sizeof(arr) / sizeof(arr[0]);

	int index = binarySearch(arr, 0, n - 1, x);

	(index == -1) ? cout << "The element has not been found." 
		      : cout << "The element is present at position " << index;
	
	return 0;
}


